gdxsv/lbs_discord.go

```diff
@@ -0,0 +1,494 @@
+		splitIndex = len(value) / (numberOfSplit - i)
+		firstHalf := value[:splitIndex]
+		splitIndex = strings.LastIndex(firstHalf, "\n")
+
+		values = append(values, value[:splitIndex])
+		// splitIndex += 2
+		value = value[splitIndex:]
+	}
+	//do not create empty fields
+	if value != "\n" {
+		values = append(values, value)
+	}
+
+	return values
+}
+
+//Embed limits: There can be up to 25 fields
+func splitEmbedFields(fields []*discordEmbedField) [][]*discordEmbedField {
+	var divided [][]*discordEmbedField
+	chunkSize := 25
+
+	for i := 0; i < len(fields); i += chunkSize {
+		end := i + chunkSize
+
+		if end > len(fields) {
+			end = len(fields)
+		}
+
+		divided = append(divided, fields[i:end])
+	}
+	return divided
+}
+
+func reducePeerStringSize(s string) string {
+	//Reduce size by removing userid
+	re := regexp.MustCompile("`.*?`\\s")
+	s = re.ReplaceAllString(s, "")
+
+	//Reduce size by replacing custom emoji (from 28 char to 1 char)
+	s = strings.ReplaceAll(s, "<:gundam:772467554160738355>", "🌎")
+	s = strings.ReplaceAll(s, "<:zaku:772467605008023563>", "🪐")
+
+	return s
+}
+
+//PublishStatusToDiscord : Update server status to the predefined Discord message thru Web Hook
+func (lbs *Lbs) PublishStatusToDiscord() {
+	if len(conf.DiscordWebhookURL) == 0 {
+		return
+	}
+
+	if atomic.CompareAndSwapUint32(&jobIsRunning, 0, 1) {
+		go func() {
+			publish(lbs)
+			atomic.StoreUint32(&jobIsRunning, 0)
+		}()
+	} else {
+		logger.Info("Request blocked!")
+		if retryTimer != nil {
+			retryTimer.Stop()
+		}
+		//Retry last blocked request after 0.5s
+		retryTimer = time.AfterFunc(time.Second/2, func() {
+			logger.Info("Retrying!")
+			publish(lbs)
+		})
+	}
+}
+
+func publish(lbs *Lbs) {
+
+	//Stop any retry request since new request appeared
+	if retryTimer != nil {
+		retryTimer.Stop()
+		retryTimer = nil
+	}
+
+	//
+	// Create battle peer list, for the third embed
+	//
+	battle := make(map[string]*battlePeers)
+
+	battlePeerCount := 0
+	var battlePeersIDs string
+	var accumulatedBattlePeersString string
+
+	for _, u := range sharedData.GetMcsUsers() {
+		_, exists := battle[u.BattleCode]
+		if exists == false {
+			battle[u.BattleCode] = new(battlePeers)
+
+			locName, ok := gcpLocationName[u.McsRegion]
+			if !ok {
+				locName = "Default Server"
+			}
+			if u.McsRegion == "best" {
+				locName = "Best Server"
+			}
+			battle[u.BattleCode].RegionName = locName
+			accumulatedBattlePeersString += locName
+		}
+
+		var peer string
+		switch u.Side {
+		case TeamRenpo:
+			peer = fmt.Sprintf("<:gundam:772467554160738355> `%s` %s\n", u.UserID, u.Name)
+			battle[u.BattleCode].RenpoPeers += peer
+		case TeamZeon:
+			peer = fmt.Sprintf("<:zaku:772467605008023563> `%s` %s\n", u.UserID, u.Name)
+			battle[u.BattleCode].ZeonPeers += peer
+		}
+		accumulatedBattlePeersString += peer
+
+		battlePeerCount++
+		battlePeersIDs += u.UserID
+	}
+
+	//
+	// Create lobby peer list, for the second embed
+	//
+	var plazaPeers string
```

Comment:
Instead of collecting information by string concatenation, let's use a data structure.

Reply:
The string concatenation is inevitable, since the `plazaUsers` string is used directly in the final Embed field's value

The current design can ensure there is only 1 single loop for `lbs.userPeers` and `sharedData.GetMcsUsers()` to create all Embed field's value

If we store the users in a data structure, we will need additional for-loop for each lobby/battle to create the concated string?

Reply:
string's `+=` is a heavy operation.
Let's use buffers.

https://hermanschaaf.com/efficient-string-concatenation-in-go/
> In Go, each time you append to the end of a string, it must create a new string and copy the contents of both the existing string and the appended string into it.

Reply:
> need additional for-loop

There is no problem with the number of for loops as long as the computational complexity does not change.
In general, the part that collects data and the part that is processed for publication should be separated.

Reply:
Updated to use buffer now